### ISSUE TITLE
OpenAI-like function calling

### ISSUE THREAD

**Issue by finom** [facebookresearch/llama#463]
Hello guys! Thank you for great job! I'm currently using OpenAI chat completions with function calling but OpenAI main models don't support fine-tuning yet but LLaMA does. Unfortunately Langchain Agents don't provide high quality of results and I'm hoping to find something that executes functions similar to OpenAI function calling (arguments, required, enum) with possibility to fine-tune. Have somebody tried to do that with LLaMA 2 (some hidden trick, some way to train to call functions etc)? Thank you!

**Comment by Chriss4123** [comment]
Fine tuning a LLM of this magnitude requires a huge amount of GPU compute. OpenAI does offer fine-tuning for models, however it is only available via the base `davinci` models. I recommend you look there first before deciding fine-tuning a custom LLM the optimal approach. Fine tuning LLaMa requires experience in autoregressive text generation and a boatload of compute.

**Comment by teleprint-me** [comment]
@finom 

I believe it's possible seeing as a few users were able to do [fine-tuning with a  A10G GPU (24 GB VRAM) in about 24 hrs](https://www.reddit.com/r/LocalLLaMA/comments/154rqay/llama2_7b_uncensored_qlora_finetune_on_wizard/) with the Wizard-Vicuna dataset.

I would assume that you would need the base model to have something like this in it's training or you could fine-tune it in a way that behaves something like the function API.

I'll be experimenting with this over the next few months or so with my local rigs.

**Comment by tytung2020** [comment]
Yes I am looking into this too. BTW, in what sense you find the LangChain agent does not provide high quality results? Did you embed the dataset for search?

**Comment by RageshAntony** [comment]
@finom 
https://www.reddit.com/r/LocalLLaMA/comments/15bkju6/bleat_function_calling_with_llama_2/

**Comment by brianjking** [comment]
There are a few fine tunes that have function calling added now.

* ToolLLM - https://github.com/OpenBMB/ToolBench
* Functionary https://github.com/MeetKai/functionary
* & this that I just came across tonight https://huggingface.co/Trelis/Llama-2-7b-chat-hf-function-calling
